filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,27 +1,115 @@
+    createBlobTokenGenerator(container, endpoint, None, None)
+  }
+  def createBlobTokenGenerator(container: String, endpoint: String, workspaceId: Option[String], workspaceManagerURL: Option[String]): BlobTokenGenerator = {
+     (container: String, endpoint: String, workspaceId, workspaceManagerURL) match {
+       case (container, endpoint, None, None) =>
+         NativeBlobTokenGenerator(container, endpoint)
+       case (container, endpoint, Some(workspaceId), Some(workspaceManagerURL)) =>
+         WSMBlobTokenGenerator(container, endpoint, workspaceId, workspaceManagerURL)
+       case _ =>
+         throw new Exception("Arguments provided do not match any available BlobTokenGenerator implementation.")
+     }
+  }
+}
+
+case class WSMBlobTokenGenerator(container: String, endpoint: String, workspaceId: String, workspaceManagerURL: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    throw new NotImplementedError
+  }
+}
+
+case class NativeBlobTokenGenerator(container: String, endpoint: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    val storageAccountName = BlobPathBuilder.parseStorageAccount(BlobPathBuilder.parseURI(endpoint)) match {
+      case Some(storageAccountName) => storageAccountName
+      case _ => throw new Exception("Storage account could not be parsed from endpoint")
+    }
+
+    val profile = new AzureProfile(AzureEnvironment.AZURE)
+    val azureCredential = new DefaultAzureCredentialBuilder()
+      .authorityHost(profile.getEnvironment.getActiveDirectoryEndpoint)
+      .build
+    val azure = AzureResourceManager.authenticate(azureCredential, profile).withDefaultSubscription
+
+    val storageAccounts = azure.storageAccounts()
+    val storageAccount = storageAccounts
+      .list()
+      .asScala
+      .find(_.name == storageAccountName)
+
+    val storageAccountKeys = storageAccount match {
+      case Some(value) => value.getKeys.asScala.map(_.value())
+      case _ => throw new Exception("Storage Account not found")
+    }
+
+    val storageAccountKey = storageAccountKeys.headOption match {
+      case Some(value) => value
+      case _ => throw new Exception("Storage Account has no keys")
+    }
+
+    val keyCredential = new StorageSharedKeyCredential(
+      storageAccountName,
+      storageAccountKey
+    )
+    val blobContainerClient = new BlobContainerClientBuilder()
+      .credential(keyCredential)
+      .endpoint(endpoint)
+      .containerName(container)
+      .buildClient()
+
+    val blobContainerSasPermission = new BlobContainerSasPermission()
+      .setReadPermission(true)
+      .setCreatePermission(true)
+      .setListPermission(true)
+    val blobServiceSasSignatureValues = new BlobServiceSasSignatureValues(
+      OffsetDateTime.now.plusDays(1),
+      blobContainerSasPermission
+    )
+
+    blobContainerClient.generateSas(blobServiceSasSignatureValues)
```

Comment:
Does it make sense for anything that feeds into `blobServiceSasSignatureValues` to be built for this `NativeBlobTokenGenerator` instance instead of every time `getAccassToken` is called? I would completely believe the answer is "no", but creating the `NativeBlobTokenGenerator` will happen once per Cromwell whereas `getAccessToken` is at least once per `BlobPathBuilder` (which, I think, translates to once per workflow).

Reply:
I like this idea! I was a little unsure saving an object with the signature values on something with a long lifetime, but I think its not really the privileged info. There is a separate question on if the permissions would have to change, but I think in the case where these are variable they are variable per Cromwell instance, and not more.